### DISCUSSION TITLE
Should we integrate theme.json design tokens into WC Blocks?

### DISCUSSION THREAD

**Discussion by danieldudzic** [woocommerce/woocommerce#44934] [WooCommerce, Blocks, and Site Editing]
I have spent quite a bit of time today trying to close the last open issue of the [TT4 + WP 6.4 compatibility project](https://github.com/woocommerce/woocommerce-blocks/issues/11201 ): https://github.com/woocommerce/woocommerce-blocks/issues/11328

@nerrad's comment made me think deeply about the subject:
> Let's think about how the borders for the design in this block should automatically adapt to what a theme might have set as [design tokens for images in the theme.json](https://github.com/WordPress/twentytwentyfour/blob/61a733f9d1050c131c3b711b7a9131f08db34b33/theme.json#L511). That way we wouldn't have theme specific CSS and the block would react more appropriately to the design of _any_ block theme.

I think it would be brilliant if somehow, we could integrate `theme.json` values and make the design of our blocks match as closely as possible the aesthetic of the theme, but this proves to be complicated.

Here are a couple of considerations:

- Currently, we are hardcoding block styles into our templates and patterns. These files are static.

- In order to try and integrate better with the theme styling, we are using spacing presets from the `theme.json`, but it turns out those don't follow naming conventions and are not uniform across themes, so we shouldn't really be using them (without a fallback): https://github.com/WordPress/twentytwentyfour/issues/710#issuecomment-1786549378).

- The "aesthetics" of the theme live in non-standardized places (when we consider multiple themes). For example in TT4, images have rounded edges. But the `border-radius` is applied to the block style variation `Rounded` and not the default, main Image block. 
<img width="738" alt="Add_New_Page_‹_newproductgallery_—_WordPress" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/f9568d21-9bc9-4e5e-8ebd-376e2131235e">

```
"core/image": {
	"variations": {
		"rounded": {
			"border": {
				"radius": "var(--wp--preset--spacing--20)"
			}
		}
	}
},
```

That said, some themes, do apply a `border-radius` to the main Image block.

In TT4 the `Quote` block, for example, has the `border-radius` added on the main block (and not a variation):

```
"core/quote": {
	"border": {
		"radius": "var(--wp--preset--spacing--20)"
	},
}

```

All of this is to say, that `theme.json` governs itself with liberty as a core principle, making it less predictable and harder to leverage in a plugin. 

---------

### Proposed solution

I was thinking, that perhaps it would be possible to create a **mapping/fallback layer** in WooCommerce Blocks.

Seems like there is an [API for hooking into the theme.json](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/filters/global-styles-filters.md).

We don't want to overwrite `theme.json` values but rather parse the styles and only apply some hand-picked values to our blocks (if they are not already styled by `theme.json`).

Sticking to the example of the border radius and images. We could define all our blocks that are/include images and would benefit from border-radius inheritance, and then map the `core/image` `border-radius` styling to them.

Style source: `data['core/image']['variations']['rounded']['border']['radius']`;

Target[1]: `data['woocommerce-blocks/product-gallery-large-image']['styles']['border']['radius'];`
Target[2]: `data['woocommerce-blocks/product-image']['styles']['border']['radius'];`
Target[3]: `data['woocommerce-blocks/featured-category-cover-image]['variations']['rounded']['border']['radius']`;

We could support multiple alternative sources to check for the presence of the border-radius (like, both style variations of the Image block), to make it compatible across multiple themes (as much as possible, at least). 

Do you think this is worthwhile exploring further?

**Comment by Aljullu** [comment]
Thanks for starting this discussion, @danieldudzic! I like the approach you proposed, technically it sounds like a nice solution and it would allow WooCommerce Blocks to adapt better to a variety of themes.

IMO, this kind of thing would ideally be provided by Gutenberg. So we could say "we want the Product Image block to inherit the global styles from the Image block". Unfortunately, that's not possible. There is the Elements API, but [it doesn't include images](https://github.com/WordPress/gutenberg/blob/f7d2103acca1f984fc5aa5f4fb8424ee6507bfe8/lib/class-wp-theme-json-gutenberg.php#L543), and it's not exactly the same.

If we go forward with this exploration, I would suggest we start small: just a few blocks and some style properties that don't break the layout. So we can revert it if needed without causing issues to existing sites.

**Comment by nerrad** [comment]
Make sure consideration is given to what `theme.json` defines as _defaults_ for styles vs those that are _options_ in the editor. `theme.json` provides configuration for both types.  For example, _style variations_ (as shared in the example of "rounded" variation) is a configuration for expressing the design option in the editor for the user to explicitly select that variation. 

What we'll want to do here is try to inherit what the theme sets as _defaults_.  I don't think we should worry about anything that describes the configuration for a design option (as in the style variations).

> If we go forward with this exploration, I would suggest we start small: just a few blocks and some style properties that don't break the layout. So we can revert it if needed without causing issues to existing sites.

Agree.
